<html>
<head>
	<link href="designgg.css" rel="stylesheet" type="text/css"/>
	<title> designGG : More information </title>
</head>

<body>
<center><table width="80%">

	<tr><td>
		<fieldset><legend align=center> Experimental Design for Genetical Genomics </legend>
		<br>
		<table>
		<tr><td>&nbsp;&nbsp;&nbsp;</td>
		<td>		
		<p align=justify>
		Genetical genomics <a href="references.html" style="text-decoration: none">[1-3]</a> is a recent strategy for studying 
		complex biological systems using multifactorial perturbation, and it has been one of the 
		most popular approaches in genomics in recent years <a href="references.html" style="text-decoration: none">[4-12]</a>. <br>
		</p>		
		<p align=justify>
		By combining molecular profiling (transcriptomics / proteomics / metabolomics / epigenomics / kinomics...) 
		and genotype data, genetical genomics has the potential to identify the functional 
		consequences of natural and induced genetic variation <a href="references.html" style="text-decoration: none">[1]</a>.<br>
		</p>		
		<p align=justify>		 
		For transcriptome data, the strategy of genetical genomics works as follows:
		</p>
		<ol> 
			<li>determine gene expression in genetically different individuals,</li> 
			<li>treat the transcript abundance of each gene in each individual as a quantitative trait,</li> 
			<li>use molecular markers to genotype the individuals,</li> 
			<li>use linkage analysis or association analysis methods to identify regulating loci (expression quantitative trait loci; eQTL), and</li> 
			<li>re-construct regulatory networks from the patterns of co-expression and eQTL co-localisation.</li>
		</ol>				
		<p align=justify>				
		Similar to this transcription-based approach, the genetic determinants of variation for protein 
		or metabolite abundance (pQTL, mQTL) or variation for any other quantitative biomolecular traits 
		(xQTL) can be mapped to genome locations and used as input for a reconstruction of biomolecular networks.<br>
		</p>		
		<p align=justify>
		Because of experimental and economical constraints, genetical genomics experiments often have to be performed
		with a very small number of samples. Therefore, it is very important to optimize the allocation of samples, 
		i.e. the experimental design, prior to performing the actual experiment.<br>
		</p>		
		<p align=justify>
		This online tool allows the user to find the optimal design for generalized genetical genomics experiments 
		[Li et al., manuscript submitted]. Given genetic information, the number of experimental factors to 
		be varied, and the number of available arrays or experimental lines (usually RILs) per condition, 
		the program will find the "best design", 
		which results in the smallest expected variance of estimates of the most relevant biological parameters.<br>
		</p>
		<p align=justify>
		More advanced settings and extra parameters of the algorithm can be adjusted for specific experimental situations.
		These are currently only available in the R version, which can be obtained from the <a href="http://www.rug.nl/staff/yang.li/index"> 
		authors</a>.  
		<br>
		</p>
		<p align=justify>
		This tool was created by <a href="http://www.rug.nl/gbic"><b>Groningen Bioinformatics Center</b></a>.<br> 
		The methods are also available as an R library, called designGG.
		</p>
		<br>	
		</td>
		<td>&nbsp;&nbsp;&nbsp;</td>
		</tr>
		</table>	
		</fieldset>	
	</td></tr>
</table></center>

</body>
</html>